pkg/search/postgres/store.go

```diff
@@ -278,13 +368,21 @@ func (s *GenericStore[T, PT]) GetMany(ctx context.Context, identifiers []string)
 func (s *GenericStore[T, PT]) DeleteByQuery(ctx context.Context, query *v1.Query) error {
 	defer s.setPostgresOperationDurationTime(time.Now(), ops.Remove)
 
-	return RunDeleteRequestForSchema(ctx, s.schema, query, s.db)
+	dbErr := RunDeleteRequestForSchema(ctx, s.schema, query, s.db)
+	s.resetCache(ctx)
```

Comment:
Do we really want to reset the cache in case an error occurred during deletion?

Also, it seems like we may want to do this within a transaction, so it might be worthwhile looking into [the sample we have](https://github.com/stackrox/stackrox/blob/0570a911c1b27210bb333bb1da899b55bf3dd946/central/blob/datastore/store/store.go#L200-L216).

Reply:
The point of reseting the cache (that is re-populating it from the database) is to ensure a stronger consistency between the cache and the underlying DB.
If it is granted that any error in the store end with no change to the persisted data set, then re-population on error is not required. Otherwise, I'd rather make sure the cache is in sync with the DB. Maybe some finer grain processing on the error received from the underlying store would prevent unnecessary refreshes (NotAllowed errors from access control would be one example).
Good point with the transactions. In that case it means we should maintain at each nested transaction level a list of removed items and a list of upserted items, merge with the next level down on commit and discard on rollback. The lowest level would be the actual cache.